### DISCUSSION TITLE
Is Gherkin being used at all?

### DISCUSSION THREAD

**Discussion by ThomasLandauer** [Codeception/Codeception#6084] [General]
As I mentioned at https://github.com/Codeception/Codeception/discussions/6080#discussion-1688364, a comparison between Behat and Codeception would be great. And I would say that https://codeception.com/docs/07-BDD is the right place for it.

But first of all I have a general question: Is Gherkin (in Codeception) really used in practice?? Or was this just some idea a couple of years ago?

What I can say in any case: You're using the term "BDD" as a synonym for Gherkin throughout the docs. I would say this is wrong, cause:
```php
$I->click('checkout');
$I->see('Total: 200 €');
```
... describes the app's **behavior** just in the same manner as:
```gherkin
When I go to checkout
Then I should see the total price is 200
```
... it's just a different syntax.
So what I'd do in any case is replace most occurrences of "BDD" with "Gherkin". After all, Codeception's description at https://github.com/Codeception/Codeception/blob/4.1/composer.json#L3 is
> BDD-style testing framework

**Comment by DavertMik** [comment]
> But first of all I have a general question: Is Gherkin (in Codeception) really used in practice?? Or was this just some idea a couple of years ago?

Yes

> When I go to checkout
Then I should see the total price is 200

No! Compare the difference

`I click "checkout"` vs `I go to checkout`

In 1st case you make an explicit action in a browser - you click a button or link with name or label `checkout`.
In 2nd case you perform something to get to checkout page - the implementation is hidden in this case. You can click a button, several buttons, open a page, do all this actions combined. You can even not use browser at all!

Each command on 1st scenario is pushed to browser to perform it.
In 2nd scenario the actual implementatoin is defined by user, so they control how they "see" or how they "go to checkout".

2nd scenario hides implementation to provide higher abstraction language. In order to focus on business processes.
1st scenario is just a test, which is simpler than scenario 2 because it doesn't have anything hidden.

**Reply by ThomasLandauer** [reply]
No, I don't agree.

Where does `I go to checkout` end up ultimately? Here:
```php
public function iGoToCheckoutProcess() {
    // 27 actions...
}
```
Which is *exactly* the same what you can do with Codeception:
```php
$I->goToCheckout();
```
...and then create a helper which is exactly the same as above, i.e. the implementation is "hidden" too.

BTW: Your answer "Yes" is to the used-in-practice question? (Not to the just-an-idea question ?)

**Reply by c33s** [reply]
yes i use gherkin

and no it's not the same. the first syntax is also more understandable for non-programmers. as programmer you are used to `$`, `;` and stuff like that. gherkin is simply more natural. it can even be written in a different language (for example german).
